eng/common/scripts/Download-Latest-Contract.ps1

```diff
@@ -0,0 +1,84 @@
+# This script downloads the project's latest GA contract if there is a GA tag provided. The contract is downloaded at {root}/../sparse-spec/sdk/{ProjectDirectory}/
+# You need to implement these functions in in Language-Settings.ps1:
+# Get-${Language}-LatestTag: Returns the tag tied with the latest GA version
+# Get-${Language}-SourceCodeSubDirectory: Returns the sub folder of source code relative to the project folder
+[CmdletBinding()]
+param (
+    [Parameter(Position=0)]
+    [ValidateNotNullOrEmpty()]
+    [string] $ProjectDirectory
+)
+
+. $PSScriptRoot/Helpers/Sparse-Clone-Helpers.ps1
+. $PSScriptRoot/common.ps1
+
+function GetCommit([string]$sdkGitRemoteAPI) {
+    try {
+        $response =  Invoke-WebRequest $sdkGitRemoteAPI
+        $responseContent = ConvertFrom-Json $([String]::new($response.Content))
+        return $responseContent.object.sha
+    }
+    catch {
+        throw $_.Exception.Message
+    }
+}
+
+function GetGitRemoteValue() {
+    Push-Location $ProjectDirectory
+    $result = ""
+    try {
+        $gitRemotes = (git remote -v)
+        foreach ($remote in $gitRemotes) {
+            if ($remote.StartsWith("origin") -and $remote -match '(?:https://github.com/|git@github.com:).*/(.*).git\s+') {
+                return $Matches[1]
+            }
+        }
+    }
+    finally {
+        Pop-Location
+    }
+
+    return $result
+}
+
+function GetProjectRelativePath() {
+    $rootPath = GetProjectRootPath $ProjectDirectory
+    $subDirectory = &$GetSourceCodeSubDirectory
+    return [System.IO.Path]::GetRelativePath($rootPath, (Join-Path $ProjectDirectory $subDirectory)).Replace("\","/")
+}
+
+if (!(Test-Path "Function:$GetLatestTagFn")) {

```

Comment:
All the languages use release tags and to determine if one is a GA or not you look at the versions. You can see an example of where I did that at https://github.com/Azure/azure-sdk/blob/main/eng/scripts/Update-Release-Versions.ps1#L201

Reply:
Seems these functions are retrieving data from https://github.com/Azure/azure-sdk/tree/main/_data/releases/latest. And it seems still need specific logic to each language to get the package name. For example, .net uses the folder name, while js uses the name indicated in package.json.

How about we just parse the `CHANGELOG.md` to get the tag? Then we don't need any cross-repo effort.

Reply:
That isn't use the csv data this process is about updating that csv data. The release tags are gathered from github in https://github.com/Azure/azure-sdk/blob/main/eng/scripts/PackageVersion-Helpers.ps1. Once you gather them you can parse and sort them to get the latest version for a given package. If getting the package name is the challenge there is already code in each language to the package information (example https://github.com/Azure/azure-sdk-for-net/blob/main/eng/scripts/Language-Settings.ps1#L13). You cannot depend on the folder name as it doesn't always match the package name. 

We also have helpers that can parse the changelog but you still need to find the right one and depending on what versions you are looking at branch it may or may not have all the correct versions in there. For example, if you are working in a feature branch that hasn't taken the latest merge from may where the last GA shipped you will miss that GA version.

Reply:
So, to get package name we still need a language specific function, correct?

And the script `PackageVersion-Helpers` calls github grapql api. That needs pat, which is not expected. We don't want our users to have such additional steps. We hope user don't need to know there is a download step.

Reply:
That example uses a PAT and calls the graphql but if you have a local clone you can also get the list of tags from the cloned repo as well.